### PR TITLE
Handle cases where 'field' keyword shouldn't generate a field

### PR DESCRIPTION
**Pull request by Youssef1313**
Test plan: https://github.com/dotnet/roslyn/issues/57012

@AlekseyTs @333fred

### CODE REVIEW

**Comment by Youssef1313** [comment]
I'm not sure if the tests timeout is related to my changes. Can someone restart to see if they still fail?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:91]: >public int P { get => field; } [](http://example.com/codeflow?start=4&length=30)

Consider also testing scenario with initializer.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:226]: >if (!CanHaveBackingField()) [](http://example.com/codeflow?start=12&length=27)

The logic here doesn't feel right to me. I think interpretation of a `field` keyword should depend only on what is in scope, not on whether containing type is interface or on the property kind. Instead, we probably should report an error that "Interfaces cannot contain instance fields", etc. Basically, report an error that would be reported for an explicit field declaration. Suppressing creation of a symbol in error scenarios doesn't look useful on its own, for example we always create a symbol when there is an initializer.

**Review by AlekseyTs — DISMISSED** [review]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Binder/FieldKeywordBinder.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/FieldKeywordBinder.cs:1]: The change here is *adding* BOM character. Visual Studio added it automatically for me.
The file was originally created in VS Code, so it seems that VS Code doesn't respect this:

https://github.com/dotnet/roslyn/blob/5e4f8781bfaa29ede55a08b978cadde3e0087ffc/.editorconfig#L15

File: src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:91]: Is that fine with a prototype with now?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:91]: > Is that fine with a prototype with now?

Can't we test this scenario right now? We don't really need the initializer to actually do any work because this is an error scenario anyway.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs:91]: >field [](http://example.com/codeflow?start=26&length=5)

We should have a test with multiple references for the same `field`. Perhaps with one in the setter too, etc.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedBackingFieldSymbol.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedBackingFieldSymbol.cs:76]: >diagnostics.Add(ErrorCode.ERR_InterfacesCantContainFields, ErrorLocation); [](http://example.com/codeflow?start=16&length=74)

It looks like this approach comes with several shortcomings:
- location is not pointing to the syntax that is the source of the problem, which is the `field` keyword in the body.
- It looks like multiple uses of the `field` keyword will report the same diagnostics at the same location. Which will appear as a bunch of duplicate errors.
- There are scenarios when this symbol is created not in response to a `field` keyword. Those scenarios are already handled and reported. Reporting more diagnostics in those scenarios is not desired. In fact it looks like a lot of existing unrelated tests are now broken because of that.

I suggest trying to report the error in Binder_Expressions.cs instead. Next to this check:
```
                            if (IsInsideNameof)
                            {
                                Error(diagnostics, ErrorCode.ERR_FieldKeywordInsideNameOf, node);
                            }
```

And report it with location of the `field` keyword.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:213]: >diagnostics [](http://example.com/codeflow?start=46&length=11)

It looks like there is a possible race here. The symbol that we create might "lose" on the next line, but it already recorded the diagnostics which is going to look like a duplicate. A "losing" symbol should have zero effect.

**Comment by AlekseyTs** [comment]
>     internal class SpeculativeFieldKeywordBinder : Binder

It doesn't look like there are any meaningful changes in this file. Please revert/exclude from the PR.

---
Refers to: src/Compilers/CSharp/Portable/Binder/SpeculativeFieldKeywordBinder.cs:12 in 6840e22. [](commit_id = 6840e22d82b4e73a955be53c8e14ce0f0dc41961, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5). Other than a couple of files without meaningful changes, LGTM.

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Need a second review on a small community PR.

**Review by 333fred — APPROVED** [review]
LGTM (commit 5)

File: src/Compilers/CSharp/Portable/Binder/FieldKeywordBinder.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/FieldKeywordBinder.cs:1]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
```

File: src/Compilers/CSharp/Portable/Binder/SpeculativeFieldKeywordBinder.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/SpeculativeFieldKeywordBinder.cs:1]:
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
```

**Review by AlekseyTs — COMMENTED** [review]
LGTM (commit 7)

**Pull request merged**
